man/chip8disasm.1
.Dd March 9, 2018
.Dt CHIP8DISASM 1
.Os
.Sh NAME
.Nm chip8disasm
.Nd disassemble Chip\-8 binaries
.Sh SYNOPSIS
.Nm
.Op Fl hqVv
.Op Fl o Ar output
.Ar file
.Sh DESCRIPTION
.Nm
is a disassembler for Chip\-8 binaries which produces output in the syntax used
by
.Xr chip8asm 1 .
It accepts one operand, the binary game file to disassemble.
The arguments are as follows:
.Bl -tag -width Ds
.It Fl h Ns , Fl \-help
Show a brief help message and exit.
.It Fl o Ar output Ns , Fl \-output Ns = Ns Ar output
Set the output file name.
If no output file name is specified or is \-, then the disassembled output will
be printed to the standard output.
.It Fl q Ns , Fl \-shift\-quirks
Enable shift quirks mode.
.It Fl V Ns , Fl \-version
Show version information and exit.
.It Fl v Ns , Fl \-verbose
Increase verbosity.
Each repetition of this flag will enable an additional level of logging output.
In order of decreasing severity, the log levels are ERROR, WARNING, INFO, DEBUG
and TRACE.
By default, only messages at the ERROR and WARNING levels are shown.
.El
.Pp
The disassembler output format is compatible with the input format used in the
assembler.
That is, if you disassemble a Chip-8 program with
.Nm
and then re-assemble it with
.Xr chip8asm 1 ,
you should get a file identical to the original.
There are some exceptions to this: for example, programs that use the
two-operand forms of
.Ic SHR
and
.Ic SHL
will need to be both disassembled and assembled in shift quirks mode.
.Ss CONTROL-FLOW ANALYSIS
One complication which the disassembler addresses is the lack of distinction
between code and data in Chip-8 assembly.
Thus, a naive disassembler that attempts to decode each 2-byte word as an
instruction will run into problems and report many invalid instructions (which
aren't really instructions at all, but data).
.Nm
solves this problem using a simple form of
.Dq control-flow analysis ,
which analyzes the program structure to detect which sections are reachable
(code) and which are not (data).
.Pp
The current implementation should work for the vast majority of programs in
existence, but isn't perfect.
The most notable omission is that the
.Ql JP V0, addr
instruction is not fully supported, so that the locations it jumps to will not
be marked as reachable.
.Ss OUTPUT LABELS
.Nm
will detect instructions in the code which reference addresses and replace the
hard-coded address values with labels to improve readability.
The labels correspond directly to the referenced addresses, disregarding the
512-byte region of memory reserved for the interpreter.
For example, the address
.Ad #4A0
corresponds to the label
.Va L2A0 .
.Sh SEE ALSO
.Xr chip8 1 ,
.Xr chip8asm 1


src/chip8disasm.c
/*
 * Copyright 2017-2018 Ian Johnson
 *
 * This is free software, distributed under the MIT license.  A copy of the
 * license can be found in the LICENSE file in the project root, or at
 * https://opensource.org/licenses/MIT.
 */
#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "disassembler.h"
#include "log.h"
#include "memory.h"

static const char *HELP =
    "A disassembler for Chip-8/Super-Chip programs.\n"
    "\n"
    "Options:\n"
    "  -o, --output=OUTPUT    set output file name\n"
    "  -q, --shift-quirks     enable shift quirks mode\n"
    "  -v, --verbose          increase verbosity\n"
    "  -h, --help             show this help message and and exit\n"
    "  -V, --version          show version information and exit\n";
static const char *USAGE = "chip8disasm [OPTION...] FILE\n";
static const char *VERSION_STRING = "chip8disasm " PROJECT_VERSION "\n";

/**
 * Options which can be passed to the program.
 */
struct progopts {
    /**
     * The output verbosity (default 0).
     *
     * The higher this is, the more log messages will be output (currently,
     * anything over 2 won't give any extra messages).
     */
    int verbosity;
    /**
     * Whether to use shift quirks mode (default false).
     */
    bool shift_quirks;
    /**
     * The output file name.
     *
     * For consistency, this should be heap-allocated memory.
     */
    char *output;
    /**
     * The input file name.
     *
     * For consistency, this should be heap-allocated memory.
     */
    char *input;
};

static struct progopts progopts_default(void);
static int run(struct progopts opts);

int main(int argc, char **argv)
{
    int option;
    struct progopts opts = progopts_default();
    const struct option options[] = {{"output", required_argument, NULL, 'o'},
        {"shift-quirks", no_argument, NULL, 'q'},
        {"verbose", no_argument, NULL, 'v'}, {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, 'V'}, {0, 0, 0, 0}};
    int retval = 0;

    log_init(argc >= 1 ? argv[0] : "chip8disasm", stderr, LOG_WARNING);

    while ((option = getopt_long(argc, argv, "o:qvhV", options, NULL)) != -1) {
        switch (option) {
        case 'o':
            free(opts.output);
            opts.output = xstrdup(optarg);
            break;
        case 'q':
            opts.shift_quirks = true;
            break;
        case 'v':
            opts.verbosity++;
            break;
        case 'h':
            printf("%s%s", USAGE, HELP);
            goto EXIT;
        case 'V':
            printf("%s", VERSION_STRING);
            goto EXIT;
        case '?':
            fprintf(stderr, "%s", USAGE);
            retval = 2;
            goto EXIT;
        }
    }

    if (optind == argc - 1) {
        opts.input = xstrdup(argv[optind]);
    } else {
        fprintf(stderr, "%s", USAGE);
        retval = 2;
        goto EXIT;
    }

    /* Use stdout as default output */
    if (!opts.output)
        opts.output = xstrdup("-");

    retval = run(opts);

EXIT:
    free(opts.output);
    free(opts.input);
    return retval;
}

static struct progopts progopts_default(void)
{
    return (struct progopts){
        .verbosity = 0, .shift_quirks = false, .input = NULL, .output = NULL};
}

static int run(struct progopts opts)
{
    struct chip8disasm *disasm;
    struct chip8disasm_options disopts = chip8disasm_options_default();
    FILE *output;
    int retval = 0;

    if (opts.verbosity == 1)
        log_set_level(LOG_INFO);
    else if (opts.verbosity == 2)
        log_set_level(LOG_DEBUG);
    else if (opts.verbosity >= 3)
        log_set_level(LOG_TRACE);

    if (opts.shift_quirks)
        disopts.shift_quirks = true;

    if ((disasm = chip8disasm_from_file(disopts, opts.input)) == NULL) {
        log_error("Could not disassemble input file '%s'", opts.input);
        retval = 1;
        goto EXIT_NOTHING_DONE;
    }
    if (strcmp(opts.output, "-") == 0) {
        output = stdout;
    } else if ((output = fopen(opts.output, "w")) == NULL) {
        log_error("Could not open output file '%s': %s", opts.output, strerror(errno));
        retval = 1;
        goto EXIT_DISASM_CREATED;
    }

    if (chip8disasm_dump(disasm, output) != 0) {
        log_error("Disassembly dump failed");
        retval = 1;
        goto EXIT_OUTPUT_OPENED;
    }

EXIT_OUTPUT_OPENED:
    if (output != stdout)
        fclose(output);
EXIT_DISASM_CREATED:
    chip8disasm_destroy(disasm);
EXIT_NOTHING_DONE:
    return retval;
}
